src/compiler/checker.ts

```diff
@@ -28417,8 +28417,8 @@ namespace ts {
 
         function checkGrammarConstructorTypeParameters(node: ConstructorDeclaration) {
             const typeParameters = getEffectiveTypeParameterDeclarations(node);
-            if (isNodeArray(typeParameters)) {
-                const { pos, end } = typeParameters;
+            if (isNodeArray(typeParameters) || typeParameters.length) {

```

Comment:
when is it not a NodeArray?

Reply:
getEffectiveTypeParameterDeclarations returns a ReadonlyArray now. Previously it returned `templateTags[0].typeParameters`, which was a NodeArray. Now it returns `flatMap(templateTags, t => t.typeParameters)`, which is a ReadonlyArray.

I don't think it's worth the trouble to create a NodeArray because the resulting span would stretch across all 3 template tags below, which is very ugly:

```js
/** @template T
* @template U
* @template V
*/
constructor() { }
```

Reply:
in the past we stuck a `pos` and an `end` on the result to make it look like a NodeArray.. why can not we do the same here?

Reply:
actually looking at the code, that is exactly what we do.. it looks like it will always be a NodeArray

Reply:
In getEffectiveTypeParameterDeclarations, the code path that calls getJSDocTypeParameterDeclarations ends up calling flatMap, which is what returns the non-NodeArray:

```ts
    /**
     * Gets the effective type parameters. If the node was parsed in a
     * JavaScript file, gets the type parameters from the `@template` tag from JSDoc.
     */
    export function getEffectiveTypeParameterDeclarations(node: DeclarationWithTypeParameters): ReadonlyArray<TypeParameterDeclaration> {
        if (isJSDocSignature(node)) {
            return emptyArray; /*** Not a NodeArray ***/
        }
        if (isJSDocTypeAlias(node)) {
            Debug.assert(node.parent.kind === SyntaxKind.JSDocComment);
            const templateTags = flatMap(filter(node.parent.tags, isJSDocTemplateTag), tag => tag.typeParameters) as ReadonlyArray<TypeParameterDeclaration>;
            const templateTagNodes = templateTags as NodeArray<TypeParameterDeclaration>;
            templateTagNodes.pos = templateTagNodes.length > 0 ? first(templateTagNodes).pos : node.pos;
            templateTagNodes.end = templateTagNodes.length > 0 ? last(templateTagNodes).end : node.end;
            templateTagNodes.hasTrailingComma = false;
            return templateTagNodes;
        }
        return node.typeParameters || (isInJavaScriptFile(node) ? getJSDocTypeParameterDeclarations(node) : emptyArray);
    }

    export function getJSDocTypeParameterDeclarations(node: DeclarationWithTypeParameters): ReadonlyArray<TypeParameterDeclaration> {
        /*** flatMap does not return a NodeArray ***/
        return flatMap(getJSDocTags(node), tag => isNonTypeAliasTemplate(tag) ? tag.typeParameters : undefined);
    }
```

The jsdoc signature case returns emptyArray, which is also not a NodeArray.